narwhal/network/src/anemo_ext.rs

```diff
@@ -44,29 +47,44 @@ impl WaitingPeer {
+        let timeout = request.timeout();
+        let sleep: OptionFuture<_> = timeout.map(tokio::time::sleep).into();
+        tokio::pin!(sleep);
         loop {
             if self.network.known_peers().get(&self.peer_id).is_none() {
                 return Err(format!("peer {} is not a known peer", self.peer_id).into());
             }
 
-            match subscriber.recv().await {
-                Ok(PeerEvent::NewPeer(peer_id)) if peer_id == self.peer_id => {
-                    // We're now connected with the peer, lets try to make a network request
-                    if let Some(mut peer) = self.network.peer(self.peer_id) {
-                        return peer.rpc(request).await.map_err(Into::into);
+            tokio::select! {
+                recv = subscriber.recv() => match recv {
+                    Ok(PeerEvent::NewPeer(peer_id)) if peer_id == self.peer_id => {
+                        // We're now connected with the peer, lets try to make a network request
+                        if let Some(mut peer) = self.network.peer(self.peer_id) {
+                            if let Some(duration) = timeout {
+                                // Reduce timeout to account for time already spent waiting
+                                // for the peer.
+                                request.set_timeout(duration.saturating_sub(Instant::now().duration_since(start)));
+                            }
+                            return peer.rpc(request).await.map_err(Into::into);
+                        }
                     }
-                }
-                Err(RecvError::Closed) => return Err("network is closed".into()),
-                Err(RecvError::Lagged(_)) => {
-                    subscriber = subscriber.resubscribe();
+                    Err(RecvError::Closed) => return Err("network is closed".into()),
+                    Err(RecvError::Lagged(_)) => {
+                        subscriber = subscriber.resubscribe();
 
-                    // We lagged behind so we may have missed the connection event
-                    if let Some(mut peer) = self.network.peer(self.peer_id) {
-                        return peer.rpc(request).await.map_err(Into::into);
+                        // We lagged behind so we may have missed the connection event
+                        if let Some(mut peer) = self.network.peer(self.peer_id) {
+                            return peer.rpc(request).await.map_err(Into::into);
+                        }
                     }
-                }
-                // Just do another iteration
-                _ => {}
+                    // Just do another iteration
+                    _ => {}
+                },
+                sleep = &mut sleep => {
+                    if sleep.is_some() {
+                        return Err(format!("timed out waiting for peer {}", self.peer_id).into());
+                    }
+                },
```

Comment:
If timeout is None then this will probably always end up resolving to None and the other branch may not get executed ever (or a lot less frequently than it should)

Reply:
```suggestion
                Some(_) = &mut sleep => {
               
                        return Err(format!("timed out waiting for peer {}", self.peer_id).into());
                   
                },
```

Reply:
Yeah that's one of the differences between futures and tokio select macros